### ISSUE TITLE
polymorphicDefault is never invoked

### ISSUE THREAD

**Issue by cy6erGn0m** [Kotlin/kotlinx.serialization#1585]
**Describe the bug**

A serializer lookup lambda registered via `polymorphicDefault` in a module is never used.

**To Reproduce**
```kotlin
@Serializable
data class X(@Polymorphic val any: Any)

private val testModule = SerializersModule {
  polymorphic(Any::class, String::class, String.serializer())
  polymorphic(Any::class, Int::class, Int.serializer())
  polymorphic(Any::class, Long::class, Long.serializer())
//  polymorphic(Any::class, Float::class, Float.serializer()) // uncomment to make it work

  polymorphicDefault(Any::class) { className ->
    when (className) {
      "kotlin.Float" -> Float.serializer() // never invoked
      else -> null
    }
  }
}

fun main() {
  val json = Json {
    serializersModule = testModule
    useArrayPolymorphism = true // the only way to have polymorphic working in json
  }

  println(json.encodeToString(X(1.0f)))
}
```

**Expected behavior**
I expect `polymorphicDefault` to be invoked in this case with serial name `kotlin.Float`.

So it should print the following instead of a crash:
```
{"any":["kotlin.Float",1.0]}
```

**Environment**
 - Kotlin version: 1.5.10
 - Library version: 1.2.1
 - Kotlin platforms: JVM

**Comment by pdvrieze** [comment]
`polymorphicDefault` is only supported for decoding, not for encoding. See:
https://github.com/Kotlin/kotlinx.serialization/blob/d489d96e84807d709cf981671457824a5b40772f/core/commonMain/src/kotlinx/serialization/modules/SerializersModule.kt#L184
and
https://github.com/Kotlin/kotlinx.serialization/blob/d489d96e84807d709cf981671457824a5b40772f/core/commonMain/src/kotlinx/serialization/modules/SerializersModule.kt#L140-L151

It is mainly designed to support handling unknown inputs from a network, not to circumvent the module system.

**Comment by sandwwraith** [comment]
Yep, it's only invoked on deserialization, but it seems there are several use-cases where it may be useful on serialization too (see similar https://github.com/Kotlin/kotlinx.serialization/issues/1317)

**Comment by cy6erGn0m** [comment]
This is inconsistent and assymmetric. Should work the same in both directions

**Comment by sandwwraith** [comment]
Duplicate of #1317 — fixed by #1686
